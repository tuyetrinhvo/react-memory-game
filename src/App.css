.App {
	text-align: center;
}

.App-logo {
	height: 40vmin;
	pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
	.App-logo {
		animation: App-logo-spin infinite 20s linear;
	}
}

.App-header {
	background-color: #282c34;
	min-height: 100vh;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	font-size: calc(10px + 2vmin);
	color: white;
}

.App-link {
	color: #61dafb;
}

@keyframes App-logo-spin {
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}

.memory {
	display: flex;
	flex-wrap: wrap;
	width: 500px;
	margin-left: auto;
	margin-right: auto;
	margin-bottom: 30px;
	user-select: none;
}

@media (max-width: 414px) {
	.memory {
		width: 400px;
	}
}

@media (max-width: 375px) {
	.memory {
    width: 350px;
    margin-left: 12px;
	}
}

@media (max-width: 320px) {
	.memory {
		width: 200px;
		margin-left: 55px;
	}
}
