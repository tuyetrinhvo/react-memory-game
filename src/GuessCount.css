.memory>.guesses {
    font-size: 0.7em;
    width: calc(100% - 0.4em);
    margin: 0.2em;
    text-align: right;
    font-family: Menlo, Monaco, Consolas, Inconsolata, 'Courier New', monospace;
}