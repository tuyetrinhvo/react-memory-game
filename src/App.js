import React, { Component } from 'react';
import './App.css';
import GuessCount from './GuessCount';
import Card from './Card';
import shuffle from 'lodash.shuffle';
import Clock from './Clock';
import HighScoreInput from './HighScoreInput';
import HallOfFrame from './HallOfFame';
import Winner from './Winner';

const SIDE = 6,
	SYMBOLS = '😀🎉💖🥰🎀😎🌼🧁😇🍀🍩🍑🥳🦋💩🧚‍♀️💍🐬👑🍄🌸🍔',
	VISUAL_PAUSE_MSECS = 750,
	END_GAME = 3000;

class App extends Component {
	// un initialiseur de champ d'instance
	// les initialiseurs sont exécutés avant tout constructeur
	// cards : un champ temporaire
	/*  cards = this.generateCards(); */

	// transforme la méthode d'instance handleCardClicked(card){}
	// en un initialiseur de champ d'instance avec une fonction fléchée
	// pour garantir conserver le this du composant App
	/*   handleCardClicked = card => {
    console.log(card, "clicked", this);
  } */

	// ou on peut mettre ces deux initialiseurs dans le constructeur
	// pour garantir conserver le this du composant App
	/*  constructor(props){
    super(props);
    this.cards = this.generateCards();
    this.handleCardClicked = card => {
      console.log(card, "clicked", this);
    }
  }
*/
	// un initialiseur de champ d'instance, et c'est un
	// Etat local, strictement privé, les autres composants ne peuvent pas y accéder
	state = {
		cards: [],
		currentPair: [],
		guesses: 0,
		matchedCardIndices: [], // liste positions des cartes déjà retournés, visible permanente
		hallOfFame: null
	};

	// Arrow function for bidding
	// on utilise l'initialiseur pour pouvoir passer par référence
	displayHallOfFrame = hallOfFame => {
		this.setState({ hallOfFame });
	};

	componentDidMount () {
		window.addEventListener('keyup', e => {
			if (e.keyCode === 13) {
				this.initGame();
			}
		});
		this.initGame();
	}

	getFeedbackForCard (index) {
		const { currentPair, matchedCardIndices } = this.state;
		const indexMatched = matchedCardIndices.includes(index); // true or false

		if (currentPair.length < 2) {
			return indexMatched || index === currentPair[0] ? 'visible' : 'hidden';
		}

		if (currentPair.includes(index)) {
			return indexMatched ? 'justMatched' : 'justMismatched';
		}

		return indexMatched ? 'visible' : 'hidden';
	}

	handleCardClicked = index => {
		const { currentPair } = this.state;

		if (currentPair.length === 2) {
			return;
		}

		if (currentPair.length === 0) {
			this.setState({ currentPair: [ index ] });
			return;
		}

		this.handleNewPairClosedBy(index);
	};

	handleNewPairClosedBy (index) {
		const { cards, currentPair, guesses, matchedCardIndices } = this.state;

		const newPair = [ currentPair[0], index ];
		const newGuesses = guesses + 1;
		const matched = cards[newPair[0]] === cards[newPair[1]];

		this.setState({ currentPair: newPair, guesses: newGuesses });

		if (matched) {
			this.setState({ matchedCardIndices: [ ...matchedCardIndices, ...newPair ] });
		}

		if (guesses === 54) {
			this.setState({ cards: [] });
			setTimeout(() => this.initGame(), END_GAME);
		}

		setTimeout(() => this.setState({ currentPair: [] }), VISUAL_PAUSE_MSECS);
	}

	generateCards () {
		const result = [],
			size = SIDE * SIDE,
			candidates = shuffle(SYMBOLS);

		while (result.length < size) {
			const card = candidates.pop();
			result.push(card, card);
		}
		return shuffle(result);
	}

	initGame () {
		this.setState({
			cards: this.generateCards(),
			currentPair: [],
			guesses: 0,
			matchedCardIndices: [],
			hallOfFame: null
		});
	}

	// index : position de card dans le tableau symbol
	render () {
		// destructure depuis this.state
		const { cards, guesses, matchedCardIndices, hallOfFame } = this.state;
		const won = matchedCardIndices.length === cards.length; // tous les cartes sont retournées de façon permanente
		return (
			<div className="memory">
				<GuessCount guesses={guesses} />
				<Clock guesses={guesses} won={won} />
				{cards.map((card, index) => (
					<Card
						card={card}
						feedback={this.getFeedbackForCard(index)}
						key={index}
						index={index}
						onclick={this.handleCardClicked}
						won={won}
					/>
				))}

				{won &&
					guesses < 55 &&
					(hallOfFame ? (
						<div>
							<Winner />
							<HallOfFrame entries={hallOfFame} />
						</div>
					) : (
						<div>
							<HighScoreInput guesses={guesses} onStored={this.displayHallOfFrame} />
						</div>
					))}
			</div>
		);
	}
}

export default App;
