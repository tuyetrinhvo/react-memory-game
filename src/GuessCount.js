import React from 'react'
import PropTypes from 'prop-types'
import './GuessCount.css'

// composant pur fonctionnel ou fonction pure
const GuessCount = ({ guesses }) => <div className="guesses">{guesses > 54 ? 
<p>Vous avez perdu ! Rejouez !</p> : <p>Vous avez tenté : {guesses} fois</p>}</div>

// pour déclarer valeur par défault, déclenchera l'avertissement dans la console si erreur
GuessCount.propTypes = {
    guesses: PropTypes.number.isRequired
}

export default GuessCount