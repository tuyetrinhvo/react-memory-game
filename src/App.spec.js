import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import App, { SYMBOLS } from './App';
import GuessCount from './GuessCount';

describe('<App />', () => {
	it('render not empty', () => {
		const wrapper = shallow(<App />);
		expect(wrapper.isEmptyRender()).to.equal(false);
	});
	it('class `memory` exists', () => {
		const wrapper = shallow(<App />);
		expect(wrapper.exists('.memory')).to.equal(true);
	});
	it('contains a zero-guess counter', () => {
		const wrapper = shallow(<App />);
		expect(wrapper).to.contain(<GuessCount guesses={0} />);
	});
	it('has 36 cards', () => {
		const wrapper = shallow(<App />);
		expect(wrapper.find('Card')).to.have.length(36);
	});
	it('should match its reference snapshot', () => {
		const mock = sinon.stub(App.prototype, 'generateCards').returns([ ...SYMBOLS.repeat(2) ]);
		try {
			const wrapper = shallow(<App />);
			expect(wrapper).to.matchSnapshot();
		} finally {
			mock.restore();
		}
	});
});
