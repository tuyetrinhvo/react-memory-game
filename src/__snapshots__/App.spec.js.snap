// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match its reference snapshot 1`] = `
<div
  className="memory"
>
  <GuessCount
    guesses={0}
  />
  <Clock
    guesses={0}
    won={false}
  />
  <Card
    card="😀"
    feedback="hidden"
    index={0}
    key="0"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🎉"
    feedback="hidden"
    index={1}
    key="1"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="💖"
    feedback="hidden"
    index={2}
    key="2"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🥰"
    feedback="hidden"
    index={3}
    key="3"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🎀"
    feedback="hidden"
    index={4}
    key="4"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="😎"
    feedback="hidden"
    index={5}
    key="5"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🌼"
    feedback="hidden"
    index={6}
    key="6"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🧁"
    feedback="hidden"
    index={7}
    key="7"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="😇"
    feedback="hidden"
    index={8}
    key="8"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍀"
    feedback="hidden"
    index={9}
    key="9"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍩"
    feedback="hidden"
    index={10}
    key="10"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍑"
    feedback="hidden"
    index={11}
    key="11"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🥳"
    feedback="hidden"
    index={12}
    key="12"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🦋"
    feedback="hidden"
    index={13}
    key="13"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="💩"
    feedback="hidden"
    index={14}
    key="14"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🧚"
    feedback="hidden"
    index={15}
    key="15"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="‍"
    feedback="hidden"
    index={16}
    key="16"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="♀"
    feedback="hidden"
    index={17}
    key="17"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="️"
    feedback="hidden"
    index={18}
    key="18"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="💍"
    feedback="hidden"
    index={19}
    key="19"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🐬"
    feedback="hidden"
    index={20}
    key="20"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="👑"
    feedback="hidden"
    index={21}
    key="21"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍄"
    feedback="hidden"
    index={22}
    key="22"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🌸"
    feedback="hidden"
    index={23}
    key="23"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍔"
    feedback="hidden"
    index={24}
    key="24"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="😀"
    feedback="hidden"
    index={25}
    key="25"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🎉"
    feedback="hidden"
    index={26}
    key="26"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="💖"
    feedback="hidden"
    index={27}
    key="27"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🥰"
    feedback="hidden"
    index={28}
    key="28"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🎀"
    feedback="hidden"
    index={29}
    key="29"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="😎"
    feedback="hidden"
    index={30}
    key="30"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🌼"
    feedback="hidden"
    index={31}
    key="31"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🧁"
    feedback="hidden"
    index={32}
    key="32"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="😇"
    feedback="hidden"
    index={33}
    key="33"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍀"
    feedback="hidden"
    index={34}
    key="34"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍩"
    feedback="hidden"
    index={35}
    key="35"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍑"
    feedback="hidden"
    index={36}
    key="36"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🥳"
    feedback="hidden"
    index={37}
    key="37"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🦋"
    feedback="hidden"
    index={38}
    key="38"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="💩"
    feedback="hidden"
    index={39}
    key="39"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🧚"
    feedback="hidden"
    index={40}
    key="40"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="‍"
    feedback="hidden"
    index={41}
    key="41"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="♀"
    feedback="hidden"
    index={42}
    key="42"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="️"
    feedback="hidden"
    index={43}
    key="43"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="💍"
    feedback="hidden"
    index={44}
    key="44"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🐬"
    feedback="hidden"
    index={45}
    key="45"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="👑"
    feedback="hidden"
    index={46}
    key="46"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍄"
    feedback="hidden"
    index={47}
    key="47"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🌸"
    feedback="hidden"
    index={48}
    key="48"
    onclick={[Function]}
    won={false}
  />
  <Card
    card="🍔"
    feedback="hidden"
    index={49}
    key="49"
    onclick={[Function]}
    won={false}
  />
</div>
`;
